Always inline functions signatures containing `f16` or `f128`

tgross35 opened the pull request:
There are a handful of tier 2 and tier 3 targets that cause a LLVM crash or linker error when generating code that contains `f16` or `f128`. The cranelift backend also does not support these types. To work around this, every function in `std` or `core` that contains these types must be marked `#[inline]` in order to avoid sending any code to the backend unless specifically requested.

However, this is inconvenient and easy to forget. Introduce a check for these types in the frontend that automatically inlines any function signatures that take or return `f16` or `f128`.

Note that this is not a perfect fix because it does not account for the types being passed by reference or as members of aggregate types, but this is sufficient for what is currently needed in the standard library.

Fixes: https://github.com/rust-lang/rust/issues/133035
Closes: https://github.com/rust-lang/rust/pull/133037

tgross35 committed:
Always inline functions signatures containing `f16` or `f128`

There are a handful of tier 2 and tier 3 targets that cause a LLVM crash
or linker error when generating code that contains `f16` or `f128`. The
cranelift backend also does not support these types. To work around
this, every function in `std` or `core` that contains these types must
be marked `#[inline]` in order to avoid sending any code to the backend
unless specifically requested.

However, this is inconvenient and easy to forget. Introduce a check for
these types in the frontend that automatically inlines any function
signatures that take or return `f16` or `f128`.

Note that this is not a perfect fix because it does not account for the
types being passed by reference or as members of aggregate types, but
this is sufficient for what is currently needed in the standard library.

Fixes: https://github.com/rust-lang/rust/issues/133035
Closes: https://github.com/rust-lang/rust/pull/133037

tgross35 commented:
I'm not sure what I am missing here - testing by setting `#[inline(always)]` the codegen tests seem to emit code (maybe a compiletest option). Aside from that, I don't think generics or traits are handled at this point.

r? @saethlin, I could use a bit of help here

saethlin commented:
You're getting code emitted because the functions are `#[no_mangle]`, which is the first thing that `cross_crate_inlinable` checks for. `contains_extern_indicator` is perhaps poorly named.

You might also run into https://github.com/rust-lang/rust/blob/c82e0dff84f922274c3060f09b5ae78d5dbf5c49/compiler/rustc_middle/src/mir/mono.rs#L107-L113 after fixing that.

RalfJung commented on compiler/rustc_mir_transform/src/cross_crate_inline.rs:
```diff
@@ -50,6 +50,16 @@ fn cross_crate_inlinable(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
         _ => {}
     }
 
+    let sig = tcx.fn_sig(def_id).instantiate_identity();
+    for ty in sig.inputs().skip_binder().iter().chain(std::iter::once(&sig.output().skip_binder()))
+    {
+        // FIXME(f16_f128): in order to avoid crashes building `core`, always inline to skip
+        // codegen if the function is not used.
```
if you move this down after `let mir`, you can just iterate over `mir.args_iter()` instead of having to call `fn_sig`

tgross35 replied:
I think this would require moving `let mir` above the checks against `opts.incremental`, `OptLevel::No`, and `threashold` to avoid hitting a `return false` before we check for the types. Would this wind up with a perf impact since we don't generate mir here in these cases?

RalfJung replied:
Hm, I thought having this check so far down would be fine. But maybe not.

saethlin replied:
Since this is a (temporary, right?) hack, I'd rather the diff be small and localized than pretty. What you've written here is good.

tgross35 replied:
> (temporary, right?)

Yes please, I don't think we will be able to stabilize these types until LLVM at least doesn't crash on everything T2+.

RalfJung commented on tests/codegen/float/f16-f128-inline.rs:
```diff
@@ -0,0 +1,67 @@
+
+pub fn f16_arg(_a: f16) {
+    // CHECK-NOT: f16_arg
+    todo!()
```
Are you sure the automatic inlining heuristics wouldn't anyway already mark these functions as inlineable? Might be worth setting cross_crate_inline_threshold to 0 to be sure that doesn't mask the test.

tgross35 replied:
I added a revision with `-Zcross-crate-inline-threshold=never -Zmir-opt-level=0 -Cno-prepopulate-passes`, is that reasonable?

saethlin replied:
No, you also need `-Copt-level=0`.

tgross35 replied:
Oh yeah of course. Done, tests still pass.

saethlin replied:
`-Cno-prepopulate-passes` just disables LLVM optimizations, rustc changes its behavior in a few places based on the value of `opt-level`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11845389524/job/33010780480) [(plain)](https://github.com/rust-lang/rust/commit/f112e9a15521eac7dc681deb4024bf64ca3f0e8b/checks/33010780480/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
...........i....ii.i.i.i...ii.................

failures:

---- [codegen] tests/codegen/float/f16-f128-inline.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll" "/checkout/tests/codegen/float/f16-f128-inline.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/float/f16-f128-inline.rs:30:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_ref_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:17:20: note: found here
; f16_f128_inline::f16_ref_arg
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:35:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_ref_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:26:20: note: found here
; f16_f128_inline::f16_ref_ret
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:40:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_ref_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:35:20: note: found here
; f16_f128_inline::f128_ref_arg
                   ^~~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:45:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_ref_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:44:20: note: found here
; f16_f128_inline::f128_ref_ret
                   ^~~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:50:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_mut_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:53:20: note: found here
; f16_f128_inline::f16_mut_arg
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:55:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_mut_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:62:20: note: found here
; f16_f128_inline::f16_mut_ret
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:60:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_mut_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:71:20: note: found here
; f16_f128_inline::f128_mut_arg
                   ^~~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:65:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_mut_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:80:20: note: found here
; f16_f128_inline::f128_mut_ret


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll
Check file: /checkout/tests/codegen/float/f16-f128-inline.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
        1: ; ModuleID = 'f16_f128_inline.59ce32246df2b6e0-cgu.0' 
        2: source_filename = "f16_f128_inline.59ce32246df2b6e0-cgu.0" 
        3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
        4: target triple = "x86_64-unknown-linux-gnu" 
        5:  
        6: @alloc_aa07815cbcb2365f7aca41cc8941a0c4 = private unnamed_addr constant <{ [19 x i8] }> <{ [19 x i8] c"not yet implemented" }>, align 1 
        7: @alloc_6bc582bffdc44de442aab6f66dd365b4 = private unnamed_addr constant <{ [48 x i8] }> <{ [48 x i8] c"/checkout/tests/codegen/float/f16-f128-inline.rs" }>, align 1 
        8: @alloc_82b2a2cbad8c28ba4564dc4ecbd7e9bc = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00\1F\00\00\00" }>, align 8 
        9: @alloc_60f1c4e6531fb6b6d90525e37dd9aa14 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00$\00\00\00" }>, align 8 
       10: @alloc_79cc007c1b095e2bdc6b96ef7d0bdf0c = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00)\00\00\00" }>, align 8 
       11: @alloc_818ee584f1ec5ee88ba376a8cbcb7c4e = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00.\00\00\00" }>, align 8 
       12: @alloc_4327bb663ca54a7b8c119379e0f05c48 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\003\00\00\00" }>, align 8 
       13: @alloc_0d04e59b3c37a8041aee1430f660b11f = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\008\00\00\00" }>, align 8 
       14: @alloc_a890b3d25fe4f6b1d177a274990e6b06 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00=\00\00\00" }>, align 8 
       15: @alloc_3c1d6e3ca54baaac102563202b3e6dbb = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00B\00\00\00" }>, align 8 
       17: ; f16_f128_inline::f16_ref_arg 
       17: ; f16_f128_inline::f16_ref_arg 
not:30                        !~~~~~~~~~~  error: no match expected
       18: ; Function Attrs: noreturn nonlazybind uwtable 
       19: define void @_ZN15f16_f128_inline11f16_ref_arg17hc34aff3c97540dd2E(ptr noalias nocapture noundef readonly align 2 dereferenceable(2) %_a) unnamed_addr #0 { 
       21: ; call core::panicking::panic 
       21: ; call core::panicking::panic 
       22:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_82b2a2cbad8c28ba4564dc4ecbd7e9bc) #2 
       23:  unreachable 
       24: } 
       25:  
       26: ; f16_f128_inline::f16_ref_ret 
not:35                        !~~~~~~~~~~  error: no match expected
       27: ; Function Attrs: noreturn nonlazybind uwtable 
       28: define noalias noundef nonnull align 2 dereferenceable(2) ptr @_ZN15f16_f128_inline11f16_ref_ret17h70d29570d32add63E() unnamed_addr #0 { 
       30: ; call core::panicking::panic 
       30: ; call core::panicking::panic 
       31:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_60f1c4e6531fb6b6d90525e37dd9aa14) #2 
       32:  unreachable 
       33: } 
       35: ; f16_f128_inline::f128_ref_arg 
       35: ; f16_f128_inline::f128_ref_arg 
not:40                        !~~~~~~~~~~~  error: no match expected
       36: ; Function Attrs: noreturn nonlazybind uwtable 
       37: define void @_ZN15f16_f128_inline12f128_ref_arg17h899b1200f7e59c3cE(ptr noalias nocapture noundef readonly align 16 dereferenceable(16) %_a) unnamed_addr #0 { 
       39: ; call core::panicking::panic 
       39: ; call core::panicking::panic 
       40:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_79cc007c1b095e2bdc6b96ef7d0bdf0c) #2 
       41:  unreachable 
       42: } 
       43:  
       44: ; f16_f128_inline::f128_ref_ret 
not:45                        !~~~~~~~~~~~  error: no match expected
       45: ; Function Attrs: noreturn nonlazybind uwtable 
       46: define noalias noundef nonnull align 16 dereferenceable(16) ptr @_ZN15f16_f128_inline12f128_ref_ret17h75f4b575024e18bcE() unnamed_addr #0 { 
       48: ; call core::panicking::panic 
       48: ; call core::panicking::panic 
       49:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_818ee584f1ec5ee88ba376a8cbcb7c4e) #2 
       50:  unreachable 
       51: } 
       52:  
       53: ; f16_f128_inline::f16_mut_arg 
not:50                        !~~~~~~~~~~  error: no match expected
       54: ; Function Attrs: noreturn nonlazybind uwtable 
       55: define void @_ZN15f16_f128_inline11f16_mut_arg17he179779c24a59fc0E(ptr noalias nocapture noundef readnone align 2 dereferenceable(2) %_a) unnamed_addr #0 { 
       57: ; call core::panicking::panic 
       57: ; call core::panicking::panic 
       58:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_4327bb663ca54a7b8c119379e0f05c48) #2 
       59:  unreachable 
       60: } 
       61:  
       62: ; f16_f128_inline::f16_mut_ret 
not:55                        !~~~~~~~~~~  error: no match expected
       63: ; Function Attrs: noreturn nonlazybind uwtable 
       64: define noalias noundef nonnull align 2 dereferenceable(2) ptr @_ZN15f16_f128_inline11f16_mut_ret17h7e5123d644701330E() unnamed_addr #0 { 
       66: ; call core::panicking::panic 
       66: ; call core::panicking::panic 
       67:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_0d04e59b3c37a8041aee1430f660b11f) #2 
       68:  unreachable 
       69: } 
       70:  
       71: ; f16_f128_inline::f128_mut_arg 
not:60                        !~~~~~~~~~~~  error: no match expected
       72: ; Function Attrs: noreturn nonlazybind uwtable 
       73: define void @_ZN15f16_f128_inline12f128_mut_arg17h2020b9428d06f7eaE(ptr noalias nocapture noundef readnone align 16 dereferenceable(16) %_a) unnamed_addr #0 { 
       75: ; call core::panicking::panic 
       75: ; call core::panicking::panic 
       76:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_a890b3d25fe4f6b1d177a274990e6b06) #2 
       77:  unreachable 
       78: } 
       79:  
       80: ; f16_f128_inline::f128_mut_ret 
not:65                        !~~~~~~~~~~~  error: no match expected
       81: ; Function Attrs: noreturn nonlazybind uwtable 
       82: define noalias noundef nonnull align 16 dereferenceable(16) ptr @_ZN15f16_f128_inline12f128_mut_ret17h512746cbe2a0f388E() unnamed_addr #0 { 
       84: ; call core::panicking::panic 
       84: ; call core::panicking::panic 
       85:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_3c1d6e3ca54baaac102563202b3e6dbb) #2 
       86:  unreachable 
       87: } 
       89: ; core::panicking::panic 
       89: ; core::panicking::panic 
       90: ; Function Attrs: cold noinline noreturn nonlazybind uwtable 
       91: declare void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1, i64 noundef, ptr noalias noundef readonly align 8 dereferenceable(24)) unnamed_addr #1 
       92:  
       93: attributes #0 = { noreturn nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       94: attributes #1 = { cold noinline noreturn nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       95: attributes #2 = { noreturn } 
       96:  
       97: !llvm.module.flags = !{!0, !1} 
       98: !llvm.ident = !{!2} 
       99:  
      100: !0 = !{i32 8, !"PIC Level", i32 2} 
      101: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
      102: !2 = !{!"rustc version 1.84.0-nightly (16bf38087 2024-11-14)"} 
------------------------------------------



```

</details>

tgross35 committed:
Pass `f16` and `f128` by value in `const_assert!`

These types are currently passed by reference, which does not avoid the
backend crashes. Change these back to being passed by value, which makes
the types easier to detect for automatic inlining.

tgross35 commented:
I need to unjunkify my commit history but as-is, this resolves the problem at https://github.com/rust-lang/rust/issues/133035.

saethlin commented:
> Should this function be updated to emit a `LocalCopy` if `generate_cgu_internal_copies || /* contains f16/f128 */`? I think this relates to Ralf's comment.

After thinking about this I think I've convinced myself that no such change is required, because that code is only run after MirUsedCollector has determined that the Instance must be codegenned. The choice of `LocalCopy` or `GloballyShared` there only changes how we're going to codegen the instance, not whether we are going to.

saethlin commented:
I've manually verified that
```
cargo +stage1 build -Zbuild-std=core --target=powerpc64-ibm-aix
cargo +stage1 build -Zbuild-std=core --target=arm64ec-pc-windows-msvc
```
Work on this PR, and not on nightly.

@bors r+

bors commented:
:pushpin: Commit 5d818914af2760972943db448d2a08de27900c96 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 5d818914af2760972943db448d2a08de27900c96 -->
<!-- homu: {"type":"Approved","sha":"5d818914af2760972943db448d2a08de27900c96","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

uweigand commented:
After this PR, building on s390x is completely broken:

```
Building stage0 library artifacts (s390x-unknown-linux-gnu)
   Compiling core v0.0.0 (/home/uweigand/rust/library/core)
   Compiling rustc-std-workspace-core v1.99.0 (/home/uweigand/rust/library/rustc-std-workspace-core)
   Compiling compiler_builtins v0.1.138
   Compiling libc v0.2.161
   Compiling alloc v0.0.0 (/home/uweigand/rust/library/alloc)
   Compiling cfg-if v1.0.0
   Compiling adler v1.0.2
rustc-LLVM ERROR: Cannot select: 0x3ff58190770: ch = store<(store (s16) into %ir.6), trunc to f16> 0x3ff580a4768, 0x3ff58190850, FrameIndex:i64<2>, undef:i64
  0x3ff58190850: f32,ch = CopyFromReg 0x3ff580a4768, Register:f32 %1
    0x3ff580bdff0: f32 = Register %1
  0x3ff581905b0: i64 = FrameIndex<2>
  0x3ff58190fc0: i64 = undef
In function: _ZN4core3f1621_$LT$impl$u20$f16$GT$5clamp8do_panic7runtime17ha09a9a0f7fcbf036E
```

It looks like the first of the two commits (pass by value) introduces the breakage.

saethlin commented:
@uweigand Can you open a new issue for this? I'm quite surprised that s390x is broken because it is a tier 2 with host tools target.
